test/integration/cert_storage_failed_test.go

```diff
@@ -0,0 +1,198 @@
+//go:build integration
+
+package integration
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"database/sql"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/sa"
+	"github.com/letsencrypt/boulder/test"
+	ocsp_helper "github.com/letsencrypt/boulder/test/ocsp/helper"
+	"golang.org/x/crypto/ocsp"
+)
+
+// getPrecertByName finds and parses a precertificate using the given hostname.
+// It returns the most recent one.
+func getPrecertByName(db *sql.DB, name string) (*x509.Certificate, error) {
+	name = sa.ReverseName(name)
+	// Find the certificate from the precertificates table. We don't know the serial so
+	// we have to look it up by name.
+	var der []byte
+	rows, err := db.Query(`
+		SELECT der
+		FROM issuedNames JOIN precertificates
+		USING (serial)
+		WHERE reversedName = ?
+	`, name)
+	for rows.Next() {
+		err = rows.Scan(&der)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if der == nil {
+		return nil, fmt.Errorf("no precertificate found for %q", name)
+	}
+
+	cert, err := x509.ParseCertificate(der)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+// expectOCSP500 queries OCSP for the given certificate and expects a 500 error.
+func expectOCSP500(cert *x509.Certificate) error {
+	_, err := ocsp_helper.Req(cert, ocsp_helper.DefaultConfig)
+	if err == nil {
+		return errors.New("Expected error getting OCSP for certificate that failed status storage")
+	}
+
+	var statusCodeError ocsp_helper.StatusCodeError
+	if !errors.As(err, &statusCodeError) {
+		return fmt.Errorf("Got wrong kind of error for OCSP. Expected status code error, got %s", err)
```

Comment:
Good point. But on the whole I prefer being able to return an error from this function and check the error rather than passing a `t *testing.T`. It feels more composable.